winsvcwrap.lua

```diff
@@ -0,0 +1,289 @@
+--[[
+Copyright 2015 Rackspace
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS-IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+--]]
+
+local table = require('table')
+local winsvc = require('winsvc')
+local winsvcaux = require('winsvcaux')
+local uv = require('uv')
+local logging = require('logging')
+local jsonStringify = require('json').stringify
+
+local exports = {}
+
+local function ReportSvcStatus(svcStatusHandle, svcStatus, dwCurrentState, dwWin32ExitCode, dwWaitHint)
+  local dwCheckPoint = 1
+
+  -- Fill in the SERVICE_STATUS structure.
+
+  svcStatus.dwCurrentState = dwCurrentState
+  svcStatus.dwWin32ExitCode = dwWin32ExitCode
+  svcStatus.dwWaitHint = dwWaitHint
+
+  if dwCurrentState == winsvc.SERVICE_START_PENDING then
+    svcStatus.dwControlsAccepted = 0
+  else
+    svcStatus.dwControlsAccepted = winsvc.SERVICE_ACCEPT_STOP
+  end
+
+  if dwCurrentState == winsvc.SERVICE_RUNNING or
+    dwCurrentState == winsvc.SERVICE_STOPPED then
+    svcStatus.dwCheckPoint = 0
+  else
+    dwCheckPoint = dwCheckPoint + 1
+    svcStatus.dwCheckPoint = dwCheckPoint
+  end
+
+  logging.infof('Report Service Status, %s', jsonStringify(svcStatus))
+  -- Report the status of the service to the SCM.
+  winsvc.SetServiceStatus(svcStatusHandle, svcStatus)
+end
+
+
+exports.SvcInstall = function(svcName, longName, desc, params)
+  local svcPath, err = winsvcaux.GetModuleFileName()
+  if svcPath == nil then
```

Comment:
i'm on the fence on this.  as written currently, its really close to the prototype service shown by the microsoft examples....  they prefer to check the actual return and then the error.